src/middlewared/middlewared/plugins/auth_/authenticate.py

```diff
@@ -58,7 +60,14 @@ async def authenticate_local_user(self, user_id, username):
 
     @private
     async def ds_authenticate(self, username, password):
-        return None  # FIXME
+        if await self.middleware.call('auth.libpam_authenticate', username, password):
+            user = await self.middleware.call('user.get_user_obj', {'username': username, 'get_groups': True})
+            return await self.common_authenticate(username, 'ds_groups', set(user['grouplist']))
+
+    @private
+    def libpam_authenticate(self, username, password):
+        p = pam.pam()
+        return p.authenticate(username, password)
```

Comment:
You should  create a `middlewared` pam file that gets passed here as an argument. This allows us to directly control whether pam_winbind, pam_ldap, and pam_kerberos are used for authentication.